source/code/plugins/VMInsightsEngine.rb

```diff
@@ -0,0 +1,323 @@
+
+        def start(config, &cb)
+            raise ArgumentError, 'config is nil' if config.nil?
+            raise ArgumentError, "config is not kind of #{Configuration}" unless config.kind_of? Configuration
+            raise ArgumentError, 'proc required' if cb.nil?
+            raise RuntimeError, 'already started' unless @thread.nil?
+            @thread = PollingThread.new config.poll_interval, config.computer, config.log, config.data_collector, cb
+            nil
+        end
+
+        def stop
+            return if @thread.nil?
+            @thread = nil if @thread.stop
+        end
+
+        def running?
+            (! @thread.nil?) && @thread.status
+        end
+
+        class Configuration
+            def initialize(computer, log, data_collector)
+                raise ArgumentError unless log
+                raise ArgumentError, "#{data_collector.class.name}" unless data_collector.kind_of? IDataCollector
+                @poll_interval = 60 # seconds
+                @computer = computer
+                @data_collector = data_collector
+                @log = log
+            end
+
+            def poll_interval=(v)
+                raise ArgumentError unless (v.kind_of? Numeric) && (v.real?) && (v >= 1)
+                @poll_interval = v
+            end
+
+            attr_reader :poll_interval, :computer, :data_collector, :log
+
+        end # class Configuration
+
+    private
+
+        class PollingThread < Thread
+            def initialize(interval, computer, log, data_collector, cb)
+                @mutex = Mutex.new
+                @condvar = ConditionVariable.new
+                @run = true
+                @log = log
+                @data_collector = data_collector
+                @cb = cb
+                @saved_exception = SavedException.new
+                @saved_cpu_exception = SavedException.new
+                @cummulative_data = CummulativeData.new
+                super() {
+                    @cummulative_data.initialize_from_baseline @data_collector.baseline
+                    MetricTuple.computer computer
+                    begin
+                        @log.info "Starting polling loop at #{interval} second interval"
+                        # try to keep as close to the polling interval, independent of
+                        # any delay waking up from the sleep, collecting data, or
+                        # emitting the message.
+                        expected_wakeup_time = Time.now
+                        while @run do
+                            expected_wakeup_time += interval
+                            now = Time.now
+                            if expected_wakeup_time <= now
+                                expected_wakeup_time = now
+                            else
+                                sleep_time = expected_wakeup_time - now
+                                @mutex.synchronize {
+                                    @condvar.wait(@mutex, sleep_time) if @run
```

Comment:
Sleeps of less than 15ms typically cannot be satisfied. In that case, you can buypass the sleep and go directly to data collection.

Reply:
So it will wake up immediately. In production the polling interval is 60 seconds. If it takes 59.985 seconds to do the data collection and queue the message, we have bigger problems.

Reply:
I don't think it will wake up immediately. It will wake up after the deadline, instead.